API xtractmime

akshaysharmajs opened the pull request:
Created main function and a function for resource metadata(still need to handle multiple content-type headers). 

Should I put all the required flags in a different file "flags.py"?

#1

AKSHAYSHARMAJS commented:
Please check these [flags or settings](https://github.com/dveselov/python-libmagic/blob/master/magic/flags.py), maybe helpful later.

elacuesta commented on main.py:
```diff
@@ -0,0 +1,40 @@
+    pass
+
+# SAMPLE INPUT
+if __name__ == '__main__':
```
Seems like this block is only for testing. In that sense, I'd suggest moving the imports from the top of the file to the block, and moving the block itself to a different file (as a temporary solution anyway, the end goal would be to have a test suite in a separate directory).

AKSHAYSHARMAJS replied:
Yes, test suite will be in a separate directory, these are just temporary.  I will move to a different file

elacuesta commented on main.py:
```diff
@@ -0,0 +1,40 @@
+
+
+#main function
+def xtractmime(body, ContentType, http_origin, NoSniff):
```
```suggestion
def xtractmime(body: bytes, content_type: List[Union[str, bytes]], http_origin: bool, no_sniff: bool) -> str:
```
* `snake_case`
* Type hints

Gallaecio replied:
I would also suggest:

-   Rename the function to something generic, e.g. `sniff_mime`, `get_mime`, `extract_mime`. `sniff_mime` may be best, since this function is the one that applies the MIME sniffing specification.

    In the future, we might expose other functions that only do a part of the specification, which could have other names (e.g. `guess_mime` for a function that only analyzes the body).

-   Requiring all parameters except for  `body` to be keyword-only. You can do this by changing `body, content_type` to `body, *, content_type`. See https://www.python.org/dev/peps/pep-3102/. I suggest this because in a function call with badly-named arguments (e.g. `xtractmime(a, b, c, d)`) the only obvious parameter is probably the body, and forcing other parameters to be specified with a keyword may improve readability (e.g. `xtractmime(a, content_type=b, http_origin=c, no_sniff=d)`).

    Also, keyword-only parameters make future API changes (e.g. deprecating a parameter) easier. When you deprecate a parameter, for a time you will log a warning if the parameter is used but your code will work as before if used. Deprecating positional parameters can be messy, because during that deprecation period the parameter in its position may be that parameter or some other parameter.

-   `content_type` → `content_types`, since it’s going to be a list.

-   It probably makes sense to set default values for `content_type`, `http_origin` and `no_sniff`.
 
    The default value for `content_type` would be an empty list. It should not be a problem in this case, since we would not edit `content_type` within the function. But let me take this opportunity to share https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html , in case you were not aware (it’s the typical thing you usually only learn after wasting a few hours debugging).

AKSHAYSHARMAJS replied:
Thank you for the suggestion, I will make the changes accordingly

elacuesta commented on main.py:
```diff
@@ -0,0 +1,40 @@
+                "text/plain; charset=iso-8859-1", 
+                "text/plain; charset=UTF-8"]
+
+check_for_apache = False
```
Should this be defined within `_get_resource_metadata`?

Gallaecio replied:
Mutable objects should not be defined at the module level most of the time, they should be defined within the function that uses them, and be passed around as function parameters if needed.

elacuesta commented on main.py:
```diff
@@ -0,0 +1,40 @@
+import requests
+from urllib.parse import urlparse 
+
+Apache_Type = ["text/plain", "text/plain; charset=ISO-8859-1", 
```
I would name this `ApacheType` or `apache_type`. Additionally, should it be "private" (underscored)?

Gallaecio replied:
Actually, being a non-mutable constant, maybe `_APACHE_TYPES` makes more sense?

elacuesta reviewed (commented):
I think it would be good to have a template for the project before doing more coding. A minimal template could consist on a package directory, `setup.py` and `tox.ini` files. https://github.com/scrapy/itemadapter/ is a fairly simple project you could look at for inspiration.

Gallaecio commented:
> Will it be better to use directly [python-magic ](https://pypi.org/project/python-magic/) or  import C header file <magic.h> just as done in python-magic[ here](https://github.com/dveselov/python-libmagic/blob/master/magic/ffi.py)?

Unless there’s a good reason to use the C file, I would use the Python package API instead.

AKSHAYSHARMAJS commented on main.py:
```diff
@@ -0,0 +1,24 @@
+check_for_apache = False
+
+#handling resource metadata
+def _get_resource_metadata(content_types, http_origin):
+    if http_origin:
+        if content_types[-1] in _APACHE_TYPES:
+            return True
+
+    return False
```
The main purpose of metadata is only for setting `check-for-apache-flag`. So, now I m returning only that. 

`_APACHE_TYPES` can contain byte sequences so should I add particular bytes in that or in a separate variable?

Gallaecio replied:
Separate variable, for sure.

AKSHAYSHARMAJS commented on main.py:
```diff
@@ -0,0 +1,24 @@
+
+
+#main function
+def extract_mime(body: bytes, *, content_types: List[Union[str, bytes]]=[], http_origin: bool=True, no_sniff: bool=False) -> str:
+    supplied_type = content_types[-1]
+
+    check_for_apache = _get_resource_metadata(supplied_type, http_origin)
+    
+    pass
```
will there be a case where content_types can be empty? Maybe handle that exception here.

AKSHAYSHARMAJS reviewed (commented):
I am working on the template will push that by tom and made some changes required please check them once.

AKSHAYSHARMAJS commented:
Please suggest what should be the value for `python-requires` and `classifiers` for `setup.py`

Gallaecio commented on setup.py:
```diff
@@ -0,0 +1,23 @@
+setup(
+	name="xtractmime",
+	version="1.0.0",
+	description="Package for sniffing MIME(Multipurpose Internet Mail Extension) types",
```
I think “package” is redundant in the description of a Python package, and I don’t think it’s worth it spelling out MIME (to those familiar it won’t help, to those unfamiliar it may confuse because of the “mail” thing).

On the other hand, I think it’s important to mention https://mimesniff.spec.whatwg.org/ , since that’s what differentiates this package from any other Python package about MIME sniffing.

What about “Implementation of the MIME Sniffing standard”?

AKSHAYSHARMAJS replied:
> What about “Implementation of the MIME Sniffing standard”?

makes more sense, also adding URL to
 "Implementation of the MIME Sniffing standard (https://mimesniff.spec.whatwg.org/)"?

Gallaecio commented on setup.py:
```diff
@@ -0,0 +1,23 @@
+	author_email="akshaysharmajs@gmail.com",
+	url="https://github.com/scrapy/xtractmime.git",
+	packages=["xtractmime"],
+	python_requires="",
+	classifiers=[
+					
+				]
```
It does not hurt to fill these out.

For `python_requires`, given there’s `--ignore-requires-python`, I think we can use 3.6, since 3.5 reached end-of-life already.

Gallaecio commented on tox.ini:
```diff
@@ -0,0 +1,16 @@
+[testenv:flake8]
+basepython = python3
+deps = 
+	flake8>=3.7.9
```
Probably makes sense to freeze flake8 as well, just as you froze `mypy` below.

Gallaecio commented on xtractmime/main.py:
```diff
@@ -0,0 +1,26 @@
+    if content_types:
+        supplied_type = content_types[-1]
+
+    check_for_apache = _get_resource_metadata(supplied_type, http_origin)
```
Why does `_get_resource_metadata` returns `check_for_apache`? Maybe the function needs a renaming.

```suggestion
    check_for_apache_bug = _should_check_for_apache_bug(supplied_type, http_origin)
```

Gallaecio reviewed (commented):
Please, remove from Git: `.DS_Store`, `.mypy_cache`, `.tox`, `xtractmime.egg-info`.

`.DS_Store` is a macOS-specific file, and you should probably add it to your [system-wide Git ignore list](https://stackoverflow.com/a/7335487/939364), because I don’t think there’s ever going to be a repository where keeping that file tracked by Git will be necessary. You can alternatively add it to your local, project specific exclude list (`.git/info/exclude`).

The rest are folders and files expected to get generated by developers working on this specific project. So they should probably be in the `.gitignore` file of the project, and that file should be committed.

Also, please consider turning your current testing code (e.g. `sample_input.py`) into automated tests. You can create a `tests/test_extract_mime.py` file, to be run with `pytest tests/`, where you can define different functions prefixes with `test_` that define a test scenario. For example:

```python
def test_empty_body_no_content_types():
    assert extract_mime(b'') == 'text/plain'
```

You can then extend this file with additional test functions as you extend the functionality of the library.

AKSHAYSHARMAJS commented on xtractmime/main.py:
```diff
@@ -0,0 +1,43 @@
+from typing import List, Union
+
+_APACHE_TYPES_STR = [
+    "text/plain",
+    "text/plain; charset=ISO-8859-1",
+    "text/plain; charset=iso-8859-1",
+    "text/plain; charset=UTF-8",
+]
+
+_APACHE_TYPES_BYTES = [
+    "746578742f706c61696e",
+    "746578742f706c61696e3b20636861727365743d49534f2d383835392d31",
+    "746578742f706c61696e3b20636861727365743d69736f2d383835392d31",
+    "746578742f706c61696e3b20636861727365743d5554462d38",
+]
```
should these be byte string or strings are fine??

AKSHAYSHARMAJS commented:
I have added the pattern matching algorithm according to section 6 and added some tests. Please check once

Also, I am thinking of storing the byte tables in a separate file `patterns.py` which can store respective byte-pattern and pattern-mask as a tupple for each type of file.
for e.g, `image_pattern = ((b"\x00\x00\x01\x00",b"\xff\xff\xff\xff"),  (b"\x00\x00\x01\x00", b"\xff\xff\xff\xff"),...)`

AKSHAYSHARMAJS commented:
Can u please confirm, I think the algo mentioned in section [6.2.1](https://mimesniff.spec.whatwg.org/#signature-for-mp4) for the matching the signature of mp4 file is wrong. 
pt 9 says `while bytes-read < box-size`.  byte-read is 16 and box-size is 4 and it will never enter while loop??

it can be `while box-size < bytes-read`??

Gallaecio commented:
> @Gallaecio Can u please confirm, I think the algo mentioned in section [6.2.1](https://mimesniff.spec.whatwg.org/#signature-for-mp4) for the matching the signature of mp4 file is wrong.
> pt 9 says `while bytes-read < box-size`. byte-read is 16 and box-size is 4 and it will never enter while loop??
> 
> it can be `while box-size < bytes-read`, if its a mistake??

I believe that the algorithm is correct, but you might have misunderstood what _box-size_ means here (e.g. you consider it to be 4?).

The first 4 bytes of the input data in MP4 files seem to indicate the size of some byte sequence they call “box”, which must contain MP4 metadata or something. The algorithm says that you must “Let box-size be the four bytes from sequence[0] to sequence[3], interpreted as a 32-bit unsigned big-endian integer”. So you are supposed to read those 4 bytes as an integer, I suspect using [struct](https://docs.python.org/3/library/struct.html). And then read the input data from the 17th byte, 4 bytes by 4 bytes, up to the box size.

AKSHAYSHARMAJS commented:
yeah, I misunderstood it 😅, Thanks for the explanation 👍🏼

Gallaecio commented on xtractmime/main.py:
```diff
@@ -0,0 +1,90 @@
+        len(pattern_mask),
+    )
+
+    assert pattern_size == mask_size, "pattern's length should match mask's length"
```
Because asserts are removed when Python code is executed in optimized mode, they should only be used to inform developers of assumptions they can make that may not be obvious. For example, here this assert would tell me that, even though it is not obvious from just reading this function, we know that this assertion will always be true because we have control over the possible combinations of `byte_pattern` and `pattern_mask` values received as input.

While this is true for the current implementation of the library, the specification opens the door for additional, user-defined table rows:

> User agents may implicitly extend this table to support additional MIME types.
> — Section 7.1, step 2

So, since we will eventually support user-defined rows, I would instead raise a `ValueError` if these sizes are not the same, which will always be raised, regardless of the Python optimization level.

Also, I’ve just realized that, although [we discussed this point](https://github.com/scrapy/xtractmime/issues/1#issuecomment-855679576), the current API does not allow for this. I believe we will need an additional parameter for the main function to allow user-defined table rows (`extra_types`?). We can leave that for last, but we should not forget about it.

AKSHAYSHARMAJS replied:
I have added the `extra_types` parameter but I will work on it later.

AKSHAYSHARMAJS replied:
Thankyou for the info, I have added `ValueError` and also a test for it.

Gallaecio commented on xtractmime/main.py:
```diff
@@ -0,0 +1,90 @@
+    if lead_whitespace:
+        while (
+            input_index < input_size
+            and bytes([input_bytes[input_index]]) in _WHITESPACE_BYTES
```
This looks overly complicated. You could turn `_WHITESPACE_BYTES` into `_WHITESPACE_BYTES = b"\t\r\x0c\n "`, and turn this into:

```suggestion
            and input_bytes[input_index] in _WHITESPACE_BYTES
```

Moreover, since the “contains” operation is the only one you need, turning `_WHITESPACE_BYTES` into a set would be best: `{b"\t", b"\r", b"\x0c", b"\n", b" "}` (or `set(b"\t\r\x0c\n ")`, for better readability).

I had read somewhere that lists could be faster for a short set of elements, but it turns out I was wrong, or at least sets seem to be much better in this specific scenario:

```python
>>> timeit.timeit(r'b" " in variable', r'variable = {b"\t", b"\r", b"\x0c", b"\n", b" "}', number=1_000_000)
0.027940084000874776
>>> timeit.timeit(r'b" " in variable', r'variable = b"\t\r\x0c\n "', number=1_000_000)
0.13945079399854876
>>> timeit.timeit(r'b"\t" in variable', r'variable = b"\t\r\x0c\n "', number=1_000_000)
0.1449901920004777
```

AKSHAYSHARMAJS replied:
`and input_bytes[input_index] in _WHITESPACE_BYTES` is failing as `input_bytes[input_index]` return the ascii value rather than return a byte value.

AKSHAYSHARMAJS commented on tests/test_main.py:
```diff
@@ -0,0 +1,77 @@
+
+    input_bytes = b"GIF87a@\x1fp\x17\xf7\x00\x00"
+
+    def test_should_check_for_apache_bug(self):
+        self.assertEqual(
+            _should_check_for_apache_bug(supplied_type="text/plain", http_origin=True),
+            True,
+        )
+        self.assertEqual(
+            _should_check_for_apache_bug(supplied_type="text/plain", http_origin=False),
+            False,
+        )
+        self.assertEqual(
+            _should_check_for_apache_bug(
+                supplied_type="application/pdf", http_origin=True
+            ),
+            False,
+        )
+        self.assertEqual(
+            _should_check_for_apache_bug(
+                supplied_type="application/pdf", http_origin=False
+            ),
+            False,
+        )
```
Thanks, it will be better to parametrize the testcases. 👍🏼

AKSHAYSHARMAJS commented on xtractmime/utils.py:
```diff
@@ -0,0 +1,82 @@
+			numbersize = parse_vint(input_bytes, input_size, Iter)
+			Iter += numbersize
+
+			if Iter < input_size - 4:
+				break
```
The test for `_is_webm_signature` in [`test_utils.py`](https://github.com/scrapy/xtractmime/pull/2/files#diff-33c13e0b177bacd2f02e29bcb8aea5b49e7ce34901fd8f41fefb65defba1bd33) is failing at above condition. Maybe I am missing something but I am following the exact algorithm for webm signature mentioned [here](https://mimesniff.spec.whatwg.org/#signature-for-webm)

Gallaecio commented on xtractmime/utils.py:
```diff
@@ -0,0 +1,82 @@
+			break
+		bytes_remain -= 1
+
+	return parsednum and numbersize
```
Since the specification does not say “Return parsed number | number size”, but “Return parsed number and number size”, I wonder if it means you should return both numbers (e.g. as a tuple). If that were the case, form what step 6.1.3 says, I would guess you would only actually need the number size. So I’m not 100% sure, but I think the right approach here may be:

```suggestion
	return numbersize
```

If I’m right here, it would also mean that you can simplify this method by removing everything about getting `parsednum`.

Note: [You are not the only one having issues with this part of the standard](https://github.com/whatwg/mimesniff/issues/93).

AKSHAYSHARMAJS replied:
I figured out that there is no need for `parsednum` only `numbersize` will be good and  `Iter >= input_size -4` this is the right condition as per this [PR](https://github.com/whatwg/mimesniff/issues/93). After these changes tests are passing for now but if something goes wrong we can change it later. ✌🏼

elacuesta commented on xtractmime/main.py:
```diff
@@ -0,0 +1,78 @@
+def extract_mime(
+    body: bytes,
+    *,
+    content_types: List[bytes] = [],
```
Please see https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments.

This also applies to `extra_types`

AKSHAYSHARMAJS replied:
@Gallaecio mentioned about this earlier [here](https://github.com/scrapy/xtractmime/pull/2#discussion_r648657242) that there is no need for it as we are not editing `content-types` and `extra_types` within the function.

elacuesta replied:
Right, in any case, I think it's better not to leave either occurrence as a list. It's not hard to handle:

`content_types: Optional[List[bytes]] = None,`

and later

`content_types = content_types or []` or `content_types = content_types if content_types is not None else []`

AKSHAYSHARMAJS replied:
Thanks, I will change them accordingly.

elacuesta commented on tests/test_utils.py:
```diff
@@ -0,0 +1,30 @@
+
+class TestUtils:
+
+    file = open("tests/files/foo.mp4","rb")
+    body_mp4 = file.read()
+    file.close()
```
```suggestion
    with open("tests/files/foo.mp4", "rb") as fp:
        body_mp4 = fp.read()
```

There are a few other occurrences.

elacuesta commented on xtractmime/main.py:
```diff
@@ -0,0 +1,78 @@
+from typing import List, Union, Tuple
```
I think it would be better to rename this file to `__init__.py`

elacuesta commented on xtractmime/main.py:
```diff
@@ -0,0 +1,78 @@
+def _is_match_mime_pattern(
+    input_bytes, byte_pattern, pattern_mask, lead_whitespace=None
+):
+    input_size, pattern_size, mask_size = (
+        len(input_bytes),
+        len(byte_pattern),
+        len(pattern_mask),
+    )
```
```suggestion
    input_size = len(input_bytes)
    pattern_size = len(byte_pattern)
    mask_size = len(pattern_mask)
```

elacuesta reviewed (commented):
Could you add type hints to the functions that are missing them (even if they're underscored). I'd suggest to add them when first writing the functions, so you don't have to go back later and inspect what the function is doing.

AKSHAYSHARMAJS commented:
> Could you add type hints to the functions that are missing them (even if they're underscored). I'd suggest to add them when first writing the functions, so you don't have to go back later and inspect what the function is doing.

Thanks for suggestion, It will be better for future reference. I will add them to every functions

Gallaecio commented on tests/test_main.py:
```diff
@@ -0,0 +1,47 @@
+import unittest
+import pytest
+
+from xtractmime.__init__ import (
+    _should_check_for_apache_bug,
+    _is_match_mime_pattern,
+    _read_resource_header,
+)
+
+from xtractmime.utils import _is_mp4_signature
```
```suggestion
import unittest

import pytest

from xtractmime import (
    _should_check_for_apache_bug,
    _is_match_mime_pattern,
    _read_resource_header,
)
from xtractmime.utils import _is_mp4_signature
```

- https://www.python.org/dev/peps/pep-0008/#imports
- https://docs.python.org/3/tutorial/modules.html#packages

Gallaecio commented on xtractmime/__init__.py:
```diff
@@ -0,0 +1,76 @@
+            input_index += 1
+
+    while pattern_index < pattern_size:
+        masked_byte = bytes([input_bytes[input_index] & pattern_mask[pattern_index]])
```
Would this work?

```suggestion
        masked_byte = input_bytes[input_index:input_index+1] & pattern_mask[pattern_index:pattern_index+1]
```

AKSHAYSHARMAJS replied:
No, it's giving this error `TypeError: unsupported operand type(s) for &: 'bytes' and 'bytes'`

Gallaecio replied:
[Looks like it can’t be helped](https://stackoverflow.com/a/22593976). I’m learning a lot :slightly_smiling_face:

Gallaecio commented on xtractmime/_patterns.py:
```diff
@@ -0,0 +1,37 @@
+
+IMAGE_PATTERN = (
```
These variables should probably be called `*_PATTERNS` (plural).

Gallaecio commented on xtractmime/utils.py:
```diff
@@ -0,0 +1,69 @@
+	if input_bytes[0:4] != b"\x1aE\xdf\xa3":
+		return False
+
+	Iter = 4
```
`Iter` is not a good name here, capital names are used mostly for classes. What about `index`?

Gallaecio commented on xtractmime/utils.py:
```diff
@@ -0,0 +1,69 @@
+def parse_vint(input_bytes, input_size, index):
+	mask = 128
+	max_vint_size = 8
+	numbersize = 1
```
Python’s standard for variable names is snake_case.

```suggestion
	number_size = 1
```

AKSHAYSHARMAJS commented:
@elacuesta should I add Type hints to testing functions??

elacuesta commented:
> @elacuesta should I add Type hints to testing functions??

No need, we don't do that for other projects. But ideally the typing checks should not pass nonetheless. If it's really a problem, some tests could be skipped [like we do in Scrapy](https://github.com/scrapy/scrapy/blob/2.5.0/setup.cfg#L25-L89).

akshaysharmajs commented:
I have just added the matching algo for mp3 without ID3 files. For testing, I can't find the mp3 file without an ID3 tag. Every file I see is ID3 tagged.

Gallaecio commented:
> I have just added the matching algo for mp3 without ID3 files. For testing, I can't find the mp3 file without an ID3 tag. Every file I see is ID3 tagged.

Maybe we can strip an MP3 from ID3 data, instead of trying to find one that does not have any?

> Also, there are a lot of problems with this algo [whatwg/mimesniff#70](https://github.com/whatwg/mimesniff/issues/70)

Bummer. I suggest you do your best, and try to come up with something that works for whatever test file you choose.

akshaysharmajs commented:
> Maybe we can strip an MP3 from ID3 data, instead of trying to find one that does not have any?

Thanks for the idea, I found some software that can strip ID3 from mp3.

akshaysharmajs commented:
> Bummer. I suggest you do your best, and try to come up with something that works for whatever test file you choose.

The whole algorithm mentioned in https://mimesniff.spec.whatwg.org/#signature-for-mp3-without-id3 is messed up and requires a lot of changes. I followed the exact algorithm mentioned in https://searchfox.org/mozilla-central/source/toolkit/components/mediasniffer/mp3sniff.c#14 and its working!! all the tests are passing 👍🏼

akshaysharmajs commented:
> > Bummer. I suggest you do your best, and try to come up with something that works for whatever test file you choose.
> 
> The whole algorithm mentioned in https://mimesniff.spec.whatwg.org/#signature-for-mp3-without-id3 is kinda confusing and requires a lot of changes. I followed the exact algorithm mentioned in https://searchfox.org/mozilla-central/source/toolkit/components/mediasniffer/mp3sniff.c#14 and its working!! all the tests are passing 👍🏼

Gallaecio commented:
> The whole algorithm mentioned in https://mimesniff.spec.whatwg.org/#signature-for-mp3-without-id3 is kinda confusing and requires a lot of changes. I followed the exact algorithm mentioned in https://searchfox.org/mozilla-central/source/toolkit/components/mediasniffer/mp3sniff.c#14 and its working!! all the tests are passing 👍🏼

Sounds good to me.

Do make sure you include a comment with links to:

- https://github.com/whatwg/mimesniff/issues/70 , to explain why the implementation does not match the standard.
- https://searchfox.org/mozilla-central/rev/f213971fbd82ada22c2c4e2072f729c3799ec563/toolkit/components/mediasniffer/mp3sniff.c , to indicate what the implementation follows.

Gallaecio commented on tests/test_utils.py:
```diff
@@ -11,24 +22,85 @@ class TestUtils:
+        ],
+    )
+    @mock.patch("xtractmime.utils._mp3_framesize")
+    def test_is_mp3_non_ID3_signature(self, mock_framesize, input_bytes, expected):
+        if input_bytes == self.body_mp3[:50]:
+            mock_framesize.return_value = 10
+            assert _is_mp3_non_ID3_signature(input_bytes) == expected
+        else:
+            mock_framesize.return_value = 417
+            assert _is_mp3_non_ID3_signature(input_bytes) == expected
```
I suggest moving the value of `mock_framesize.return_value` to `@pytest.mark.parametrize`, even if it means repeating it 3 times.

Gallaecio reviewed (commented):
You should probably also rename `utils.py` to `_utils.py`, unless users are meant to be able to use any object there. And that way you no longer need to use `_` before object names defined there.

akshaysharmajs commented:
There are some tests for which flake8 and black are conflicting. Should I add them to ignore list??

Gallaecio commented:
> There are some tests for which flake8 and black are conflicting.

```
xtractmime/_patterns.py:16:100: E501 line too long (105 > 99 characters)
xtractmime/_patterns.py:30:100: E501 line too long (100 > 99 characters)
xtractmime/_patterns.py:41:100: E501 line too long (120 > 99 characters)
xtractmime/_patterns.py:51:100: E501 line too long (100 > 99 characters)
xtractmime/_patterns.py:56:100: E501 line too long (150 > 99 characters)
xtractmime/_patterns.py:57:100: E501 line too long (156 > 99 characters)
setup.py:12:100: E501 line too long (100 > 99 characters)
```

This is caused by long strings. [There’s a way around that](https://docs.python.org/3/reference/lexical_analysis.html#string-literal-concatenation):

```python
FONT_PATTERNS = (
    (
        (
            b"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
            b"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00LP"
        ),
        (
            b"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
            b"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\xff"
        ),
        "application/vnd.ms-fontobject",
    ),  # 34 bytes followed by the string "LP", the Embedded OpenType signature
)
```

[This one it seems we should add to the allow-list](https://github.com/psf/black/issues/280):

```
xtractmime/__init__.py:47:40: E203 whitespace before ':'
xtractmime/__init__.py:53:53: E203 whitespace before ':'
xtractmime/_utils.py:67:34: E203 whitespace before ':'
xtractmime/_utils.py:99:29: E203 whitespace before ':'
xtractmime/_utils.py:111:33: E203 whitespace before ':'
xtractmime/_utils.py:123:26: E203 whitespace before ':'
```

I’m not sure of my personal preference, but that’s what black is for, to avoid most endless discussions about personal preferences :slightly_smiling_face: .

```
xtractmime/__init__.py:74:5: F841 local variable 'check_for_apache' is assigned to but never used
xtractmime/__init__.py:75:5: F841 local variable 'resource_header' is assigned to but never used
tests/test_utils.py:1:1: F401 'unittest' imported but unused
tests/test_main.py:1:1: F401 'unittest' imported but unused
```

Those are code issues. Remove the unused code.

```
tests/test_utils.py:135:47: E711 comparison to None should be 'if cond is None:'
tests/test_utils.py:139:46: E711 comparison to None should be 'if cond is None:'
tests/test_utils.py:143:49: E711 comparison to None should be 'if cond is None:'
```

black is not forcing your hand here either. You can just replace `==` with `is`, and flake8 will stop complaining, while black will not revert your change.

Gallaecio commented on xtractmime/__init__.py:
```diff
@@ -1 +1,77 @@
+    no_sniff: bool = False,
+    extra_types: Optional[List[Tuple[Union[str, bytes]]]] = None,
+) -> str:
+    content_types = content_types if content_types is not None else []
+    extra_types = extra_types if extra_types is not None else []
+    supplied_type = content_types[-1] if content_types else None
```
I don’t think you need the first line. And I think the second line can be simplified, to improve readability.

```suggestion
    extra_types = extra_types or tuple()
    supplied_type = content_types[-1] if content_types else None
```

akshaysharmajs commented:
> [This one it seems we should add to the allow-list](https://github.com/psf/black/issues/280):
> 
> ```
> xtractmime/__init__.py:47:40: E203 whitespace before ':'
> xtractmime/__init__.py:53:53: E203 whitespace before ':'
> xtractmime/_utils.py:67:34: E203 whitespace before ':'
> xtractmime/_utils.py:99:29: E203 whitespace before ':'
> xtractmime/_utils.py:111:33: E203 whitespace before ':'
> xtractmime/_utils.py:123:26: E203 whitespace before ':'
> ```

We should also add [`xtractmime/__init__.py:47:13: W503 line break before binary operator`](https://github.com/psf/black/issues/113) to the list

Gallaecio commented on setup.py:
```diff
@@ -1,9 +1,20 @@
-    version="0.0.0",
+    version="1.0.0",
+    license="BSD",
+    description="""Implementation of the MIME Sniffing standard 
+                   (https://mimesniff.spec.whatwg.org/)""",
```
This will introduce unintended extra spaces. Use this instead:

```suggestion
    description=(
        "Implementation of the MIME Sniffing standard  (https://mimesniff.spec.whatwg.org/)"
    ),
```

Gallaecio commented on xtractmime/_patterns.py:
```diff
@@ -0,0 +1,102 @@
+        b"\xff\xff\xff\xff\x00\x00\x00\x00\xff\xff\xff\xff\xff\xff",
+        "image/webp",
+    ),  # The string "RIFF" followed by four bytes followed by the string "WEBPVP"
+    (
+        b"\x89PNG\r\n\x1a\n",
+        b"\xff\xff\xff\xff\xff\xff\xff\xff",
+        "image/png",
+    ),  # An error-checking byte followed by the string "PNG" followed by CR LF SUB LF,
+    # the PNG signature
```
When comments are for code that either spans multiple lines or is too long, they should go before the code they are for, not inline in the last line of the code.

```suggestion
    # An error-checking byte followed by the string "PNG" followed by CR LF SUB LF,
    # the PNG signature
    (
        b"\x89PNG\r\n\x1a\n",
        b"\xff\xff\xff\xff\xff\xff\xff\xff",
        "image/png",
    ),
```

Gallaecio commented on xtractmime/_patterns.py:
```diff
@@ -38,23 +40,28 @@
     (
-        b"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00LP",
-        b"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\xff",
+        b"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
+        b"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00LP",
+        b"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00"
+        b"\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\xff",
```
I suggested using parentheses here for readability. Without them, you need to check whether or not there is a comma at the end, while with parentheses it’s more obvious. I believe black won’t remove the parentheses.

Gallaecio commented on xtractmime/_patterns.py:
```diff
@@ -0,0 +1,102 @@
+IMAGE_PATTERNS = (
```
Maybe you could include a comment with a reference to the section of the standard where these tables are defined, so that it’s clear why these variables are sorted the way they are, i.e. not alphabetically.

For example:

```suggestion
# 6.1, step 1
# https://mimesniff.spec.whatwg.org/commit-snapshots/609a3a3c935fbb805b46cf3d90768d695a1dcff2/#matching-an-image-type-pattern
IMAGE_PATTERNS = (
```

This also leads me to something else: we are working with a live standard, which means it is not versioned, it keeps changing. So when including links in the documentation and code, it’s best to point to the specific snapshot of the standard that the documentation or code is referring to. That way, if the documentation or code does not match the standard in the future, it will be easier to tell whether it was a bug on our side or the fact that the standard changed.

On the living standard document you can find a link titled “Snapshot as of this commit” which will lead you to the latest snapshot of the standard. Please, use that link instead of a link to the living standard document whenever you link to the standard for reference. The only exception I can think of would be when mentioning the standard itself, instead of specific content of it, such as in the README or in the package description.

Gallaecio commented on xtractmime/_utils.py:
```diff
@@ -0,0 +1,252 @@
+]
+
+
+def _is_mp4_signature(input_bytes: bytes) -> bool:
```
Now that it’s `_utils`, you no longer need to keep these function names private as well.

```suggestion
def is_mp4_signature(input_bytes: bytes) -> bool:
```

Gallaecio commented on xtractmime/_utils.py:
```diff
@@ -0,0 +1,252 @@
+    ARCHIVE_PATTERNS,
+)
+
+sample_rates = [44100, 48000, 32000]
```
These values do not need to be mutable, so a tuple makes more sense. And non-mutable module-level variables are conceptually constants, which should be uppercased.

```suggestion
SAMPLE_RATES = (44100, 48000, 32000)
```

Gallaecio commented on xtractmime/_utils.py:
```diff
@@ -0,0 +1,252 @@
+
+
+def _is_mp3_non_ID3_signature(input_bytes: bytes) -> bool:
+    """This implementation does not match with standards due to various problems with the
+    algorithm according to https://github.com/whatwg/mimesniff/issues/70.
+
+    The current implementation follows
+    https://dxr.mozilla.org/mozilla-central/source/toolkit/components/mediasniffer/mp3sniff.c
+    as the algorithm for MP3 without ID3 sniffing mentioned in standards is originally
+    based on mp3sniff.c."""
```
The first paragraph should always say what the function does. The fact that it does not match the implementation should come later. So start with something like “Return True if the input bytes belong to an MP3 file without ID3 metadata, or False otherwise.”

Gallaecio commented on xtractmime/_utils.py:
```diff
@@ -0,0 +1,252 @@
+
+    box_size = unpack(">I", input_bytes[0:4])[0]
+
+    if (input_size < box_size) or (box_size % 4 != 0):
+        return False
```
```suggestion
    if input_size < box_size or box_size % 4:
        return False
```

Gallaecio commented on xtractmime/_utils.py:
```diff
@@ -0,0 +1,252 @@
+    return False
+
+
+def _parse_vint(input_bytes: bytes, input_size: int, index: int) -> int:
```
Given the implementation change, this may be a more appropriate name:

```suggestion
def parse_vint_number_size(input_bytes: bytes, input_size: int, index: int) -> int:
```

I’m also thinking that a single `input_bytes` parameter would make more sense, provided that call `_parse_vint(input_bytes, input_size, index)` gets replaced by something like `_parse_vint(memoryview(input_bytes)[index:])`. In the implementation, `index` will always be 0, 
and `len(input_bytes)` should be a valid replacement for `input_size`, and I think it’s actually fixing the algorithm. This also applies to similar functions below; if the algorithm can be implemented with just a memoryview as input, it’s probably better to do that.

Finally, in these functions I think it would also make sense to precede them with a code comment with a link to the snapshot of the standard they are modeled from, for future reference. And, in this specific case, you could explain why the difference with the standard.

akshaysharmajs replied:
I think you are talking about the `_match_mp3_header` function under `_is_mp3_non_ID3_signature` where index is initially 0 but it changes later for the second call of the function. 

In case of `_parse_vint`, there too  index keeps on changing. Also, I think it would not be good to use `memoryview` for the _parse_vint function because of the bitwise AND operation associated with it in `_parse_vint` as below.
```
def parse_vint_number_size(input_bytes: bytes, index: int) -> int:
    input_size = len(input_bytes)
    mask = 128
    max_vint_size = 8
    number_size = 1

    while number_size < max_vint_size and number_size < input_size:
        if input_bytes[index] & mask != 0:
```

Gallaecio replied:
> think it would not be good to use `memoryview` for the _parse_vint function because of the bitwise AND operation associated with it in `_parse_vint` as below.

Could you elaborate why? `bytes` and `memoryview` seem to work the same as far as bitwise operators is concerned:

```python
>>> value = b'\x79\x80'
>>> value_reference = memoryview(value)
>>> value[0] & 128
0
>>> value_reference[0] & 128
0
>>> value[1] & 128
128
>>> value_reference[1] & 128
128
```

akshaysharmajs replied:
Sorry, the memoryview is working. I was testing it wrong earlier 😅

akshaysharmajs replied:
So, the function is definition looks like `def parse_vint_number_size(input_bytes: memoryview, index: int) -> int:` now and the function call will be `parse_vint_number_size(memoryview(input_bytes), index)`

Gallaecio replied:
My suggestion is to remove the `index` parameter as well.

If you build the memory view as `memoryview(input_bytes)[index:]`, then you can pass the memory view _only_ as a parameter, and assume the `index` within the function to be `0`.

akshaysharmajs replied:
> If you build the memory view as memoryview(input_bytes)[index:], then you can pass the memory view only as a parameter, and assume the index within the function to be 0.

I believe that will change the `input_size`. Will that be a problem??

Gallaecio replied:
I might be wrong, but given `max_vint_size` is `8`, and given input_size is only used in `while number_size < max_vint_size and number_size < input_size:`, I think the goal of `input_size` is preventing going beyond the length of `input_bytes` (i.e. do not let `input_bytes[index]` raise an `IndexError`).

So I think setting `input_size` to `len(input_bytes)` without the function, if `input_bytes` in the function comes from `memoryview(input_bytes)[index:]` outside the function, would still do the job.

If I understand the algorithm correctly, this is what it does (in a more Pythonic way):

```python
    mask = 128
    max_vint_size = 8
    limit = min(max_vint_size, input_size)
    first_byte = input_bytes[0]
    for number_size in range(1, limit):
        if first_byte & mask:
            return number_size
        mask >>= 1
    return limit
```

I’m not 100% sure, so I suggest you try to think of a scenario where the results would be different before and after this change, and we can discuss whether that scenario proves that the suggested change breaks thing, or that there was a minor bug in the initial implementation.

akshaysharmajs replied:
you are right, the above changes are working, I tried all possible scenarios 👍🏼

Gallaecio commented on xtractmime/_utils.py:
```diff
@@ -0,0 +1,252 @@
+    if input_size < 4:
+        return False
+
+    if input_bytes[0:4] != b"\x1aE\xdf\xa3":
+        return False
```
```suggestion
    if input_bytes[:4] != b"\x1aE\xdf\xa3":
        return False
```

Gallaecio commented on xtractmime/_utils.py:
```diff
@@ -0,0 +1,252 @@
+        return False
+
+    index = 4
+
+    while index < input_size and index < 38:
```
```suggestion
    limit = min(input_size, 38)
    while index < limit:
```

Gallaecio commented on xtractmime/_utils.py:
```diff
@@ -0,0 +1,252 @@
+
+    if (
+        input_bytes[index : index + 1] != b"\xff"
+        or bytes([input_bytes[index + 1] & 224]) != b"\xe0"
```
According to the standard:

```suggestion
        and bytes([input_bytes[index + 1] & 224]) != b"\xe0"
```

I strongly suggest adding a test for this; a test that would have failed with the current implementation, but passed after this fix.

akshaysharmajs replied:
The algorithm mentioned in standards is wrong it should be `or` according to the `is_mp3` function below mentioned in mozzila implementation
```
static int is_mp3(const uint8_t* p, long length) {
 /* Do we have enough room to see a 4 byte header? */
      if (length < 4) return 0;
      /* Do we have a sync pattern? */
      if (p[0] == 0xff && (p[1] & 0xe0) == 0xe0) {
        /* Do we have any illegal field values? */
        if (((p[1] & 0x06) >> 1) == 0) return 0;  /* No layer 4 */
        if (((p[2] & 0xf0) >> 4) == 15) return 0; /* Bitrate can't be 1111 */
        if (((p[2] & 0x0c) >> 2) == 3) return 0;  /* Samplerate can't be 11 */
        /* Looks like a header. */
        if ((4 - ((p[1] & 0x06) >> 1)) != 3) return 0; /* Only want level 3 */
        return 1;
      }
  return 0;
}
```

Gallaecio reviewed (commented):
You seem to be going really well, so please keep it up!

Since we seem to be going a bit ahead of schedule, I’ll take the chance to bomb you with style feedback :slightly_smiling_face:

akshaysharmajs commented:
Thank you so much for the reviews and suggestions. I am learning a lot of new things and will try my best to work on them 😄

Gallaecio commented:
I just had a though for a stretch goal: automate the generation of `_pattern.py`. It’s not necessary for project completion, but it would be great since it would remove the chance of human error while transcribing and make it trivial to update the file as they add new patterns to the standard in the future.

akshaysharmajs commented:
As per the discussion, I will open a new PR with the implementation of section 7 to merge into this branch `highlvl-API` just to make things less complicated.

Gallaecio commented on xtractmime/_patterns.py:
```diff
@@ -1,102 +1,156 @@
+        b"\xff\xff\xff\xff\xff\xff",
+        None,
+        "image/gif",
+    ),  # The string "GIF87a", a GIF signature
```
The position of this type of comment affects many entries of the file that used to be a single line but Black turned into multiple lines. Black is not smart enough to fix that.

akshaysharmajs replied:
Maybe it will be better to remove all the comments for each table entry as we are already mentioning the snapshot of the table of patterns

Gallaecio commented on xtractmime/_patterns.py:
```diff
@@ -1,102 +1,156 @@
+"""
+Section 6.1, step 1
+https://mimesniff.spec.whatwg.org/commit-snapshots
+/609a3a3c935fbb805b46cf3d90768d695a1dcff2/#matching-an-image-type-pattern
+"""
```
Please, use a comment. Module-level docstrings are meant to document the module itself.

If these module-level constants where to go into Sphinx documentation, you could use `#:` instead of `#` for the comments, and Sphinx would be able to pick up the comments as documentation for the following module-level object. But since that’s not the case here, a simple comment should suffice.

Also, [it’s unusual to split URLs, even if they surpass the line length guideline](https://stackoverflow.com/a/10740010/939364). I think static analyzers (i.e. black and flake8) don’t complain in these cases; but it they do, it’s better to add an inline flag to ignore these specific errors than to split URLs. This applies to `_utils.py` as well.

Gallaecio commented on xtractmime/_utils.py:
```diff
@@ -46,14 +46,14 @@
 ]
```
`MP25_RATES` and `MP3_RATES` should probably be tuples as well, for the same reason as `SAMPLE_RATES`.

Gallaecio reviewed (commented):
I’m trying to figure out why the Windows tests are failing, but it’s a bit hard to navigate the test output because of the use of long parameters (file bodies) in `parametrize`.

I suggest you use the file name in the parameters (e.g. `foo.gif`), and move the file-reading code into the test functions themselves.

For example:

```python
    @pytest.mark.parametrize(
        "input_bytes,expected",
        [
            ("foo_mp4", True),
            (b"\x00\x00\x00", False),
        ],
    )
    def test_is_mp4_signature(self, input_bytes, expected):
        if isinstance(input_bytes, str):
            with open(f"tests/files/{input_bytes}", "rb") as input_file:
                input_bytes = input_file.read()
        assert is_mp4_signature(input_bytes) == expected
```

Gallaecio commented:
OK, I think the length of parametrize parameters is actually what’s causing the Windows errors. Pytests seems to try and put the parameters (as part of the test name) into an environment variable, and Windows does not support environment variables that long:

```
2021-06-30T19:27:04.8104969Z self = environ({'NUMBER_OF_PROCESSORS': '2', 'COMSPEC': 'C:\\Windows\\system32\\cmd.exe', 'PROCESSOR_ARCHITECTURE': 'AMD64', ...V_CORE_SOURCE': 'xtractmime', 'COV_CORE_CONFIG': ';', 'COV_CORE_DATAFILE': 'D:\\a\\xtractmime\\xtractmime\\.coverage'})
2021-06-30T19:27:04.8106169Z key = 'PYTEST_CURRENT_TEST'
2021-06-30T19:27:04.8107430Z value = 'tests/test_utils.py::TestUtils::test_is_mp4_signature[\\x00\\x00\\x00 ftypmp42\\x00\\x00\\x00\\x00mp42mp41isomavc1\\x...ZZZ]\\xe0\\x82\\x14\\xb4\\xb4\\xb4\\xb4\\xb4\\xb4\\xb4\\xb4\\xb4\\xb4\\xb4\\xb4\\xb4\\xb4\\xb4\\xb4\\xbc-True] (setup)'
2021-06-30T19:27:04.8108504Z 
2021-06-30T19:27:04.8108945Z     def __setitem__(self, key, value):
2021-06-30T19:27:04.8109481Z         key = self.encodekey(key)
2021-06-30T19:27:04.8110076Z         value = self.encodevalue(value)
2021-06-30T19:27:04.8110658Z >       self.putenv(key, value)
2021-06-30T19:27:04.8111370Z E       ValueError: the environment variable is longer than 32767 characters
2021-06-30T19:27:04.8111867Z 
2021-06-30T19:27:04.8112444Z c:\hostedtoolcache\windows\python\3.8.10\x64\lib\os.py:681: ValueError
```

akshaysharmajs commented:
> OK, I think the length of parametrize parameters is actually what’s causing the Windows errors. Pytests seems to try and put the parameters (as part of the test name) into an environment variable, and Windows does not support environment variables that long:

So should I strip the body of diff files into smaller number of bytes like `body_mp4[:1000]` or something like that?

Gallaecio commented:
Regarding contexts: I think it would make sense to add an additional, enum parameter to the main function to allow users to select a context, and have it be the web browser context by default. That said, I would treat that as an stretch goal, and I would even prioritize automating the `_patterns.py` generation over it.

elacuesta commented on tests/requirements.txt:
```diff
@@ -0,0 +1,3 @@
+dataclasses; python_version == "3.6"
```
I'm not sure this file is necessary

elacuesta commented on setup.cfg:
```diff
@@ -1,2 +1,3 @@
 [flake8]
+ignore = E203, W503
```
Can we have specific `# noqa (E203)` comments at each affected line? I'm worried this will cover other cases, like the one in [this link](https://www.flake8rules.com/rules/E203.html) (note that specifying the code in the comment is optional, just to be extra clear about what it's supposed to be ignored, but "noqa" ignores all checks).

elacuesta commented on tox.ini:
```diff
@@ -37,4 +37,4 @@ deps =
 commands =
     mypy --show-error-codes --ignore-missing-imports \
-        --follow-imports=skip {posargs: xtractmime setup.py tests}
+        --follow-imports=skip {posargs: xtractmime setup.py tests}
```
```suggestion
        --follow-imports=skip {posargs: xtractmime setup.py tests}
        
```
This is a no-change, the suggestion will remove it from the PR.

elacuesta commented on xtractmime/_utils.py:
```diff
@@ -0,0 +1,269 @@
+
+
+def parse_vint_number_size(input_bytes: memoryview) -> int:
+    """Returns an integer value by which the index in the current input bytes of a
```
(Nitpick alert) two suggestions about this and the other docstrings in the file, from [PEP 257 -- Docstring Conventions](https://www.python.org/dev/peps/pep-0257/):
* "The docstring is a phrase ending in a period. It prescribes the function or method's effect as a command ("Do this", "Return that"), not as a description; e.g. don't write "Returns the pathname ..."."
* "Unless the entire docstring fits on a line, place the closing quotes on a line by themselves"

akshaysharmajs commented:
Should I make the changes for the type of input and return values in this PR or in another one?
The structure of  `extract_mime` will be 
```
def extract_mime(
    body: bytes,
    *,
    content_types: Optional[Tuple[Union[bytes, str]]] = None,
    http_origin: bool = True,
    no_sniff: bool = False,
    extra_types: Optional[Tuple[Tuple[bytes, bytes, Set[bytes], str], ...]] = None,
) -> bytes:
```

Also, I will have to change all the comparison with string mimetype to bytes and all the fourth column values in `_patterns.py` 
to bytes. Same goes with all the type hints.

Maybe a check for string type in `extract_mime` will be better.
```
if type(supplied_type) == str:
    supplied_type = supplied_type.encode()
```

Gallaecio commented on xtractmime/__init__.py:
```diff
@@ -42,14 +45,10 @@ def _is_match_mime_pattern(
     body: bytes,
     *,
-    content_types: Optional[Tuple[bytes]] = None,
+    content_types: Optional[Tuple[Union[bytes, str]]] = None,
```
Didn’t we agree to make it only `bytes` in our last meeting?

akshaysharmajs replied:
That's what I mentioned in my last comment above 😅. I think we agreed on input type to be (bytes, str) and output to be only bytes but I might be wrong.

Gallaecio replied:
@elacuesta Do you remember?

elacuesta replied:
Not sure to be honest, but accepting only bytes will most likely simplify things.

elacuesta commented on setup.py:
```diff
@@ -1,9 +1,21 @@
 setuptools.setup(
     name="xtractmime",
-    version="0.0.0",
+    version="1.0.0",
```
```suggestion
    version="0.0.0",
```

This version should be modified in tandem with the one in .bumpversion.cfg (ideally by using the `bumpversion` CLI tool, but it's technically not mandatory). Let's hold on incrementing it until we make a release.

elacuesta commented on xtractmime/_utils.py:
```diff
@@ -0,0 +1,272 @@
+        return False
+
+
+def is_image(input_bytes: bytes) -> Optional[bytes]:
```
As with `is_mp3_non_ID3_signature`, I would expect functions starting with `is_` to be boolean predicates. Could you comment on this, maybe adding docstrings to this and the following functions?

Gallaecio replied:
Or/and rename them.

elacuesta replied:
Right, I meant commenting here on the naming decision, not as in "adding comments" :sweat_smile:

akshaysharmajs replied:
I have already changed the names of the functions in other PR. Should I make those changes here too?

Gallaecio replied:
Right. I guess it’s better to merge as-is, then, otherwise things can get messy when merging.

Gallaecio commented:
Going back to @elacuesta ’s https://github.com/scrapy/xtractmime/pull/2#discussion_r663850053, I believe that issue will be handled by black, so I think it would be OK to ignore the error code globally instead of line-by-line, which requires us to add the error code to every slice at the moment.

Gallaecio commented:
@akshaysharmajs Since this needs to be merged before #2, let us know which remaining issues you will fix as part of #4, fix the rest here, and let us merge this soon.

akshaysharmajs commented:
For this PR, I have fixed:
- the version
- global E203
- `extract_mime` type hints
- fail-fast: false

rest of the issues will be fixed as a part of #4 including the naming of functions
